### PR TITLE
Add spilled data in query summary, Web UI

### PR DESCRIPTION
**Pull request by anoopj**
This commit adds instrumentation about the spilled data size in the
query summary and the web UI.

This implementation adds an operator stat that is rolled up
to the query level. I followed the approach in the comment:
https://github.com/prestodb/presto/pull/8010#issuecomment-314844081

Note that I have not included the React-generated query JavaScript. I can
include that also.

### CODE REVIEW

**Comment by anoopj** [comment]
The [Travis build](https://travis-ci.org/prestodb/presto/jobs/454338268) failed because I did not include the React-generated JavaScript in the pull request. Should I include that too?

```
ERROR: Generated dist folder did not match checked-in version
Refer to the root README.md for instructions on generating Web UI
The command "if [[ -v WEBUI_CHECKS ]]; then
  presto-main/bin/validate_webui_generated_files.sh
fi
" exited with 1.
```

**Comment by findepi** [comment]
@anoopj yes, you need to commit also re-generated files. The idea is that we change UI not that very often and we don't want to increase build times for everyone because of the UI.

cc @raghavsethi

**Comment by anoopj** [comment]
I've included the React-generated JavaScript also and the tests have [passed](https://travis-ci.org/prestodb/presto/builds/455097045). Please let me know if there is any feedback.

**Comment by raghavsethi** [comment]
@electrum do we want to make spill part of StatementStats?

**Comment by anoopj** [comment]
*Friendly bump* 

@electrum @raghavsethi

**Comment by raghavsethi** [comment]
@anoopj I'm pretty sure we don't want spill info shown in the common case (when there is no spilling happening). You should make those changes to the user-visible bits. I'll wait for @electrum to confirm on the client API.

File: presto-client/src/main/java/com/facebook/presto/client/StatementStats.java [hunk 1]
```diff
@@ -44,6 +45,7 @@
     private final long processedBytes;
     private final long peakMemoryBytes;
     private final StageStats rootStage;
+    private final DataSize spilledDataSize;
```
Comment (electrum) [inline presto-client/src/main/java/com/facebook/presto/client/StatementStats.java]: Let's move this up to be directly after `peakMemoryBytes`, since it is related and should be next to the other memory stats. Put the field, constructor argument, assignment, getter, and `toString`, all in the same order. And update the builder as well.

Also, while in general it's good to use `DataSize`, we should be consistent with the other stats in this class. So, please change it to `long spilledBytes`.

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java [hunk 2]
```diff
@@ -442,6 +442,7 @@ public OperatorStats getOperatorStats()
                 outputPositions.getTotalCount(),
 
                 succinctBytes(physicalWrittenDataSize.get()),
+                succinctBytes(spillContext.getSpilledBytes()),
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java]: Let's move this down to be after the `peak` stats. Also, surround it with blank lines since it is a separate stat.

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java [hunk 3]
```diff
@@ -540,6 +543,11 @@ public String toString()
                     .add("usedBytes", reservedBytes.get())
                     .toString();
         }
+
+        public long getSpilledBytes()
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java]: Move this up to be after `updateBytes`

File: presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java [hunk 4]
```diff
@@ -57,6 +57,7 @@
     private final long outputPositions;
 
     private final DataSize physicalWrittenDataSize;
+    private final DataSize spilledDataSize;
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/operator/OperatorStats.java]: Move this down to be after the `peak` stats (here and everywhere else in this class)

File: presto-main/src/test/java/com/facebook/presto/execution/TestQueryStats.java [hunk 5]
```diff
@@ -57,10 +57,11 @@
                     succinctBytes(116L),
                     117L,
                     succinctBytes(118L),
-                    new Duration(119, NANOSECONDS),
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/execution/TestQueryStats.java]: Changing these unrelated values makes the diff hard to read. Use a new number for the new value rather than trying to keep the values sequential. The only change here should be adding the new value.

File: presto-main/src/test/java/com/facebook/presto/execution/TestQueryStats.java [hunk 6]
```diff
@@ -262,5 +265,6 @@ public static void assertExpectedQueryStats(QueryStats actual)
 
         assertEquals(420, actual.getWrittenPositions());
         assertEquals(58, actual.getLogicalWrittenDataSize().toBytes());
+        assertEquals(657, actual.getSpilledDataSize().toBytes());
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/execution/TestQueryStats.java]: We should move this to be after the `peak` stats.

**Comment by electrum** [comment]
I agree with @raghavsethi that it's probably cleaner to only print the value in the CLI and UI if it is greater than zero. Though for "physical written bytes" (and possibly others), we always show it in the UI, even when it's not relevant.

**Comment by anoopj** [comment]
@electrum thanks for the review. I agree with all of your feedback. Will send a new patch.

**Comment by anoopj** [comment]
Sorry about the merge commit - I can send another PR if you would like.

**Comment by raghavsethi** [comment]
You can pull-rebase and force push this branch to update the PR. You don't need to create a new one.

**Comment by anoopj** [comment]
Rebased and updated the pull request with all the feedback incorporated. Tested the CLI and the UI and verified.

File: presto-client/src/main/java/com/facebook/presto/client/StatementStats.java (hunk 1 above)
Comment (raghavsethi) [reply presto-client/src/main/java/com/facebook/presto/client/StatementStats.java]: @anoopj : Does not look like this is resolved. Please fix this.

Comment (anoopj) [reply presto-client/src/main/java/com/facebook/presto/client/StatementStats.java]: @raghavsethi 

I believe I fixed it. Are you looking at the [latest diff](https://github.com/prestodb/presto/pull/11910/files#diff-5d764bf71dd84521d4d24b7256b536b2)? Please let me know if I've missed anything.

Comment (raghavsethi) [reply presto-client/src/main/java/com/facebook/presto/client/StatementStats.java]: I can still see `spilledDataSize` when I look at the commit.
screen shot 2018-12-20 at 2 43 37 pm

Comment (anoopj) [reply presto-client/src/main/java/com/facebook/presto/client/StatementStats.java]: @raghavsethi Sorry, I missed the part about renaming the variable. I had only changed the type, not the name. I've fixed it. Please let me know if there is any other feedback.

File: presto-main/src/test/java/com/facebook/presto/execution/TestQueryStats.java [hunk 7]
```diff
@@ -58,16 +58,17 @@
                     succinctBytes(116L),
                     117L,
                     succinctBytes(118L),
-                    new Duration(119, NANOSECONDS),
-                    120L,
-                    new Duration(121, NANOSECONDS),
+                    new Duration(120, NANOSECONDS),
```
Comment (raghavsethi) [inline presto-main/src/test/java/com/facebook/presto/execution/TestQueryStats.java]: This change seems unrelated. Please just add the new line

Comment (anoopj) [reply presto-main/src/test/java/com/facebook/presto/execution/TestQueryStats.java]: Done.  Please review.

**Comment by anoopj** [comment]
@raghavsethi All requested changes done. Please review.

**Comment by raghavsethi** [comment]
I will merge this with minor edits as soon as release window opens up again.

**Comment by raghavsethi** [comment]
Merged, thank you for your contribution!

**Comment by tooptoop4** [comment]
i can't see spilled on the UI in prestodb 0.220 ?
